cli/task_runner.rs

```diff
@@ -0,0 +1,506 @@
+      &mut env_vars,
+      node_modules_dir.join(".bin").to_string_lossy().to_string(),
+    );
+  }
+  env_vars
+}
+
+fn prepend_to_path(env_vars: &mut HashMap<String, String>, value: String) {
+  match env_vars.get_mut("PATH") {
+    Some(path) => {
+      if path.is_empty() {
+        *path = value;
+      } else {
+        *path =
+          format!("{}{}{}", value, if cfg!(windows) { ";" } else { ":" }, path);
+      }
+    }
+    None => {
+      env_vars.insert("PATH".to_string(), value);
+    }
+  }
+}
+
+pub fn real_env_vars() -> HashMap<String, String> {
+  std::env::vars()
+    .map(|(k, v)| {
+      if cfg!(windows) {
+        (k.to_uppercase(), v)
+      } else {
+        (k, v)
+      }
+    })
+    .collect::<HashMap<String, String>>()
+}
+
+// WARNING: Do not depend on this env var in user code. It's not stable API.
+pub(crate) const USE_PKG_JSON_HIDDEN_ENV_VAR_NAME: &str =
+  "DENO_INTERNAL_TASK_USE_PKG_JSON";
+
+pub struct NpmCommand;
+
+impl ShellCommand for NpmCommand {
+  fn execute(
+    &self,
+    mut context: ShellCommandContext,
+  ) -> LocalBoxFuture<'static, ExecuteResult> {
+    if context.args.first().map(|s| s.as_str()) == Some("run")
+      && context.args.len() > 2
+      // for now, don't run any npm scripts that have a flag because
+      // we don't handle stuff like `--workspaces` properly
+      && !context.args.iter().any(|s| s.starts_with('-'))
+    {
+      // run with deno task instead
+      let mut args = Vec::with_capacity(context.args.len());
+      args.push("task".to_string());
+      args.extend(context.args.iter().skip(1).cloned());
+
+      let mut state = context.state;
+      state.apply_env_var(USE_PKG_JSON_HIDDEN_ENV_VAR_NAME, "1");
+      return ExecutableCommand::new(
+        "deno".to_string(),
+        std::env::current_exe().unwrap(),
+      )
+      .execute(ShellCommandContext {
+        args,
+        state,
+        ..context
+      });
+    }
+
+    // fallback to running the real npm command
+    let npm_path = match context.state.resolve_command_path("npm") {
+      Ok(path) => path,
+      Err(err) => {
+        let _ = context.stderr.write_line(&format!("{}", err));
+        return Box::pin(futures::future::ready(
+          ExecuteResult::from_exit_code(err.exit_code()),
+        ));
+      }
+    };
+    ExecutableCommand::new("npm".to_string(), npm_path).execute(context)
+  }
+}
+
+pub struct NodeCommand;
+
+impl ShellCommand for NodeCommand {
+  fn execute(
+    &self,
+    context: ShellCommandContext,
+  ) -> LocalBoxFuture<'static, ExecuteResult> {
+    // run with deno if it's a simple invocation, fall back to node
+    // if there are extra flags
+    let mut args = Vec::with_capacity(context.args.len());
+    if context.args.len() > 1
+      && (
+        context.args[0].starts_with('-') // has a flag
+        || !matches!(
+          MediaType::from_str(&context.args[0]),
+          MediaType::Cjs | MediaType::Mjs | MediaType::JavaScript
+        )
```

Comment:
I feel like there might already be some code like this somewhere, but didn't immediately see it

Reply:
This seems ok.